tests/test_resolver.py

```diff
@@ -73,6 +73,48 @@ def test_resolvable_set():
     rs.merge(rq, [binary_pkg])
 
 
+def test_resolvable_set_is_constraint_only():
+  builder = ResolverOptionsBuilder()
+  rs = _ResolvableSet()
+  c = ResolvableRequirement.from_string('foo', builder)
+  c.is_constraint = True
+
+  package = SourcePackage.from_href('foo-2.3.4.tar.gz')
+  rs.merge(c, [package])
+
+  assert rs.packages() == [(c, set([package]), None, True)]
+
+
+def test_resolvable_set_constraint_and_non_constraint():
+  builder = ResolverOptionsBuilder()
+  rs = _ResolvableSet()
+  constraint = ResolvableRequirement.from_string('foo', builder)
+  constraint.is_constraint = True
```

Comment:
maybe I'm missing something, but I'd expect to see a test here that:

1) adds a constraint for `foo==$SOME_VERSION`.
2) adds a non-constraint requirement for `foo`.
3) given multiple visible `foo` versions (some higher, some lower than $SOME_VERSION), selects the appropriate one based on the constraint.

and then maybe another variant of that which doesn't apply the constraint and would end up with the very latest version of foo available under the same conditions (sans the constraint)?

seem reasonable?